doc/man/man8/slapdn.8
.TH SLAPDN 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 2004-2011 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP: pkg/ldap/doc/man/man8/slapdn.8,v 1.6.2.11 2010/04/13 20:22:46 kurt Exp $
.SH NAME
slapdn \- Check a list of string-represented LDAP DNs based on schema syntax
.SH SYNOPSIS
.B SBINDIR/slapdn
[\c
.BI \-d \ debug-level\fR]
[\c
.BI \-f \ slapd.conf\fR]
[\c
.BI \-F \ confdir\fR]
[\c
.BR \-N | \-P ]
[\c
.BI \-o \ option\fR[ = value\fR]]
[\c
.BR \-v ]
.IR DN \ [...]
.LP
.SH DESCRIPTION
.LP
.B Slapdn
is used to check the conformance of a DN based on the schema
defined in
.BR slapd (8)
and that loaded via 
.BR slapd.conf (5).
It opens the
.BR slapd.conf (5)
configuration file or the slapd\-config (5) backend, reads in the schema definitions, and then
parses the 
.I DN
list given on the command-line.
.LP
.SH OPTIONS
.TP
.BI \-d \ debug-level
enable debugging messages as defined by the specified
.IR debug-level ;
see
.BR slapd (8)
for details.
.TP
.BI \-f \ slapd.conf
specify an alternative
.BR slapd.conf (5)
file.
.TP
.BI \-F \ confdir
specify a config directory.
If both
.B \-f
and
.B \-F
are specified, the config file will be read and converted to
config directory format and written to the specified directory.
If neither option is specified, an attempt to read the
default config directory will be made before trying to use the default
config file. If a valid config directory exists then the
default config file is ignored.
.TP
.BI \-N
only output a normalized form of the \fIDN\fP, suitable to be used
in a normalization tool; incompatible with
.BR \-P .
.TP
.BI \-o \ option\fR[ = value\fR]
Specify an
.I option
with a(n optional)
.IR value .
Possible generic options/values are:
.LP
.nf
              syslog=<subsystems>  (see `\-s' in slapd(8))
              syslog\-level=<level> (see `\-S' in slapd(8))
              syslog\-user=<user>   (see `\-l' in slapd(8))

.fi
.TP
.BI \-P
only output a prettified form of the \fIDN\fP, suitable to be used
in a check and beautification tool; incompatible with
.BR \-N .
.TP
.B \-v
enable verbose mode.
.SH EXAMPLES
To check a
.B DN
give the command:
.LP
.nf
.ft tt
	SBINDIR/slapdn \-f /ETCDIR/slapd.conf \-v DN
.ft
.fi
.SH "SEE ALSO"
.BR ldap (3),
.BR slapd (8),
.BR slaptest (8)
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.27 Release
	Build environment
		Fix getopt(3) detection

OpenLDAP 2.0.26 Release
	Fixed slapd ACL buffer size bug (ITS#1963)
	Fixed slapd ACL filters w/substrings bug (ITS#2023, ITS#2027)
	Fxied slapd directory string space bug (ITS#2072)
	Fixed slurpd reject file handling bugs (ITS#2007, ITS#2009)
	Fixed back-ldbm indirect IDL bug (ITS#2039)
	Fixed libldap buffer size bugs (ITS#1935, ITS#1936)
	Fixed libldap gethostbyname mutex init bug (ITS#2019)
	Fixed liblutil passwd bug (ITS#2055)
	Fixed sasl large reads and writes (ITS#1983, also ITS#910,ITS#1861)
	Updated back-ldbm reindexing - slapindex rebuilds dn2id index
	Build environment
		Require version 1.5 if Cyrus SASL is present
		Remove tempnam checking and replacment implementation

OpenLDAP 2.0.25 Release
	Fixed slapd extended op reqdata crash bug (ITS#1721)

OpenLDAP 2.0.24 Release
	Fixed slapd max incoming macro bug (ITS#1828)
	Fixed slapd acl group/dnaddr bug (ITS#1607)
	Fixed slapd connection management bug (ITS#1655)
	Fixed slapd port ntoh / addr init bugs (ITS#1615, ITS#1728)
	Fixed slapd PF_lOCAL typo (ITS#1660)
	Fixed slapd extended ops with NULL reqdata bug (ITS#1721)
	Fixed slapd replog logging without replica (ITS#1335)
	Fixed slapd NULL attrval indexing (ITS#1843)
	Fixed slurpd tls init bug (ITS#1613)
	Fixed back-ldbm dn normalization bug in onelevel searches (ITS#1654)
	Fixed back-ldbm modrdn root dn check (ITS#1761)
	Fixed back-ldbm idl overrun bug (ITS#1570)
	Fixed back-ldbm passwd textlen bug (ITS#1742)
	Fixed ber_bvstrdup() empty string bug (ITS#1662)
	Fixed ldapsearch filter bug (ITS#1649)
	Fixed maildap NULL domain bug (ITS#1022)
	Fixed libldap hostless URL bug (ITS#1697)
	Fixed liblutil /dev/urandom incomplete reads (ITS#1834)
	Fixed libldap_r threadpool free list (ITS#1839)
	Fixed libldap_r NT threads cond_broadcast (ITS#1865)
	Fixed sasl writes > 64K (ITS#1861)
	Fixed sasl session callback corruption (ITS#1799)
	Build environment
		Removed lint (ITS#1744)
	Documentation
		Updated release documents

OpenLDAP 2.0.23 Release
	Fixed slapd moddn acl bug (ITS#1562) (ITS#1582)
	Fixed slapd acl regex (to dn="") portability bug (ITS#1579)
	Disabled -lldap atexit() handler (ITS#1577)
	Build environment
		Updated config.guess/sub scripts (ITS#1567)
	Documentation
		Fixed a number of typos (ITS#1578)

OpenLDAP 2.0.22 Release
	Fixed slapd passwd modify referral bug
	Fixed back-ldbm index threading bug
	Fixed back-ldbm ordering presence index bug
	Fixed back-ldap modify bug (ITS#1547)
	Fixed -lldap TLS errno bug
	Fixed ud v2 kbind bug
	Updated slapd slurpd friendly code
	Build environment
		Fixed Mach cthreads detection (ITS#1554)
		Updated Berkeley DB 4 detection

OpenLDAP 2.0.21 Release
	Fix slapd empty string indexing/filtering handling (ITS#1507)
	Changed slapd ACL selfwrite replace logic (ITS#1530)
	Build environment
		Updated IPv6 configuration argument handling
		Fix back-shell tools make depend bug (ITS#1518)
		Removed extraneous files

OpenLDAP 2.0.20 Release
	Fixed slapd ACL empty replace bug (ITS#1530)
	Fixed slapd ACL peername/sockname exact match bug (ITS#1516)
	Fixed back-passwd db_config bug
	Fixed -lldap cache debug bug (ITS#1501)
	Fixed -lldap dnssrv free bug
	Fixed -lldap Start TLS first op bug
	Fixed slurpd TLS non-critical/critical bug (ITS#1527)
	Fixed slurpd SASL password bug (ITS#1528)
	Build environment
		Fixed repl_user build error (ITS#1503)
		Updated BerkeleyDB 4 support
		Updated schema files
		Updated Windows NT/2000 MSVC5 port

OpenLDAP 2.0.19 Release
	Fixed back-ldbm not filter indexing (ITS#1405)
	Fixed back-passwd e_ndn bug (ITS#1467)
	Fixed back-ldap undef ad bug (ITS#1367)
	Fixed slapd suffix "" bugs (ITS#1430)
	Fixed slapd zero length replace bug (ITS#1364)
	Fixed slapd teletex string printables bug (ITS#1419)
	Fixed slapd unallocated sockets bug (ITS#1415)
	Fixed slapd undefined present filter bug (ITS#1439)
	Fixed -lldap cacertfile bug (ITS#1406)
	Fixed -lldap ldap_url_seach_s bug (ITS#1429)
	Fixed -lldap unbind free bug (ITS#1442)
	Fixed -llber realloc bug (ITS#1410)
	Fixed ldappasswd/slappasswd strncmp bug (ITS#1497)
	Fixed various memory leaks
	Updated slapd emfile/enfile handling
	Build environment
		Fixed passwd-search port bug (ITS#1461)
		Fixed Kerberos build bug (ITS#1468)
		Updated CThread detection (ITS#1440)
		Updated MAXPATHLEN handling (ITS#1440)
		Fixed TLS build (ITS#1412)
		Added BerkeleyDB 4 support
		Added GNU ar support (ITS#1460)

OpenLDAP 2.0.18 Release
	Fixed -llber ber_bvecadd bug (ITS#1398)
	Fixed slap tools replica config bug (ITS#1385)
	Build environment
		Updated req_query detection

OpenLDAP 2.0.17 Release
	Fixed ldbm BerkeleyDB2 compat bug (ITS#1390)

OpenLDAP 2.0.16 Release
	Fixed slapd indexing bug (ITS#1349)
	Fixed slapd/slurpd maxargs bug (ITS#1343)
	Fixed slapd ManageDSAIT warning
	Fixed slapd no OID schema bug
	Fixed slurpd server down reject bug (ITS#1183)
	Fixed slurpd moddn newsup bug (ITS#1379)
	Fixed -llber ber_realloc bug (ITS#1346)
	Fixed -lldap multi-level referral chase bug (ITS#1346)
	Fixed -lldap high port bug (ITS#1375)
	Fixed CR/LF handling (ITS#1328)
	Updated slapd/-lldap getaddrinfo handling (ITS#1304)
	Updated -lldap_r pthread stack size
	Updated slapd acl assert checks
	Updated slurpd ManageDSAIT handling
	Updated slapd root/subschema DSE update error handling
	Removed -lldbm use of Berkeley CDB
	Updated -lldbm to use r/w lock with DB_THREAD
	Updated back-ldbm update attribute indices on delete (ITS#1359)
	Updated ud Kerberos ticket handling
	Added back-ldbm idl_check() code
	Build environment
		Updated ltconfig for MacOS X 10.1 (ITS#1361)
		Updated slapd.conf with ACL example
		Updated pthread detection
		Updated Kerberos detection (ITS#1372)
		Updated test005 /dev/null diff for Tru64 compat
		Updated test007 modrdn test
		Updated test008 read/write ratio
		Updated MSVC5 project files
	Documentation
		Update slapcat/add pages

OpenLDAP 2.0.15 Release
	Fixed -lldap TLS external handling

servers/slapd/slapdn.c
/* $OpenLDAP: pkg/ldap/servers/slapd/slapdn.c,v 1.8.2.5 2010/04/13 20:23:21 kurt Exp $ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 2004-2011 The OpenLDAP Foundation.
 * Portions Copyright 2004 Pierangelo Masarati.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Pierangelo Masarati for inclusion
 * in OpenLDAP Software.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/string.h>
#include <ac/socket.h>
#include <ac/unistd.h>

#include <lber.h>
#include <ldif.h>
#include <lutil.h>

#include "slapcommon.h"

int
slapdn( int argc, char **argv )
{
	int			rc = 0;
	const char		*progname = "slapdn";

	slap_tool_init( progname, SLAPDN, argc, argv );

	argv = &argv[ optind ];
	argc -= optind;

	for ( ; argc--; argv++ ) {
		struct berval	dn,
				pdn = BER_BVNULL,
				ndn = BER_BVNULL;

		ber_str2bv( argv[ 0 ], 0, 0, &dn );

		switch ( dn_mode ) {
		case SLAP_TOOL_LDAPDN_PRETTY:
			rc = dnPretty( NULL, &dn, &pdn, NULL );
			break;

		case SLAP_TOOL_LDAPDN_NORMAL:
			rc = dnNormalize( 0, NULL, NULL, &dn, &ndn, NULL );
			break;

		default:
			rc = dnPrettyNormal( NULL, &dn, &pdn, &ndn, NULL );
			break;
		}

		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "DN: <%s> check failed %d (%s)\n",
					dn.bv_val, rc,
					ldap_err2string( rc ) );
			if ( !continuemode ) {
				rc = -1;
				break;
			}
			
		} else {
			switch ( dn_mode ) {
			case SLAP_TOOL_LDAPDN_PRETTY:
				printf( "%s\n", pdn.bv_val );
				break;

			case SLAP_TOOL_LDAPDN_NORMAL:
				printf( "%s\n", ndn.bv_val );
				break;

			default:
				printf( "DN: <%s> check succeeded\n"
						"normalized: <%s>\n"
						"pretty:     <%s>\n",
						dn.bv_val,
						ndn.bv_val, pdn.bv_val );
				break;
			}

			ch_free( ndn.bv_val );
			ch_free( pdn.bv_val );
		}
	}
	
	if ( slap_tool_destroy())
		rc = EXIT_FAILURE;

	return rc;
}
